### PR TITLE
JAX (non-jit) uses the `JacobianProductCalculator` and device provided vjps

### PR DESCRIPTION
**Pull request by albi3ro**
[sc-39465]

This PR ports over the jax (non-jit) interface to use the jacobian product calculator, and creates a logical pathway to bind vjps with jax instead of jvps.  The vjp pathway would then only be used when we want to bind device vjps to jax.

Note that we cannot yet support device vjps and jax jacobians, as jax adds a broadcast dimension to `dy` that default qubit is not yet capable of handling.

PR #4854  has show  we are not quite ready to integrate our PyTree definitions with how we interact with jax.  It currently has some remaining issues:

* Not all operators are actually valid pytrees, see `ParametrizedEvolution`, `ParametrizedHamiltonian`, `HardwareHamiltonian`
* Validation done on initialization prevents the operators from being used to store cotangents
* What starts out as a vanilla numpy array can be turned into a jax tracer by the boundary

Instead of reverting that PR to still manually strip out and manually repack trainable parameters, I decided to close that PR and open a fresh one.

### CODE REVIEW

File: pennylane/interfaces/jax.py [hunk 1]
```diff
@@ -15,24 +15,47 @@
...
 logger = logging.getLogger(__name__)
 logger.addHandler(logging.NullHandler())
 
 
+Batch = Tuple[qml.tape.QuantumTape]
+ExecuteFn = Callable[[Batch], qml.typing.ResultBatch]
+
+
+@dataclasses.dataclass
+class NonPytreeWrapper:
```
Comment (trbromley) [inline pennylane/interfaces/jax.py:42]: Do we need this visible in the docs? If this is something that only internal PennyLane developers are expected to see then please could we not have it in the docs.

Comment (albi3ro) [reply pennylane/interfaces/jax.py]: did a documentation pass on the draft.

File: pennylane/ops/functions/dot.py [hunk 2]
```diff
@@ -89,18 +89,21 @@ def dot(
     # When casting a Hamiltonian to a Sum, we also cast its inner Tensors to Prods
     ops = [qml.prod(*op.obs) if isinstance(op, Tensor) else op for op in ops]
 
-    if coeffs[0] != 1 and qml.math.allequal(coeffs[0], coeffs):
+    if not qml.math.allclose(coeffs[0], 1) and qml.math.allequal(coeffs[0], coeffs):
```
Comment (timmysilv) [inline pennylane/ops/functions/dot.py:92]: do you know what caused this inaccuracy?

File: tests/interfaces/default_qubit_2_integration/test_jax_default_qubit_2.py [hunk 3]
```diff
@@ -787,6 +826,7 @@ def test_multiple_hamiltonians_not_trainable(
 
         res = jax.jacobian(cost_fn)(weights, coeffs1, coeffs2)
         expected = self.cost_fn_jacobian(weights, coeffs1, coeffs2)[:, :2]
+        print(res, expected)
```
Comment (timmysilv) [inline tests/interfaces/default_qubit_2_integration/test_jax_default_qubit_2.py:829]:
```suggestion
```

File: tests/interfaces/test_jax.py [hunk 4]
```diff
@@ -146,20 +146,21 @@ def cost(params):
 
         # adjoint method only performs a single device execution per tape, but gets both result and gradient
         assert dev.num_executions == 3
-        spy.assert_called()
+        spy.assert_not_called()
```
Comment (timmysilv) [inline tests/interfaces/test_jax.py:149]: what caused this change?

**Review by timmysilv — COMMENTED** [review]
looks great! just a few questions/curiosities

File: pennylane/interfaces/execution.py [hunk 5]
```diff
@@ -43,7 +44,7 @@
 
 device_type = Union[qml.Device, "qml.devices.Device"]
 
-jpc_interfaces = {"autograd", "numpy", "torch", "pytorch"}
+jpc_interfaces = {"autograd", "numpy", "torch", "pytorch", "jax", "jax-python"}
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/execution.py:47]: created a technical backlog story for "interface names are standardized".

File: pennylane/interfaces/execution.py [hunk 6]
```diff
@@ -778,7 +787,14 @@ def device_gradient_fn(inner_tapes, **gradient_kwargs):
             execute_fn = partial(ml_boundary_execute, execute_fn=execute_fn, jpc=jpc)
             jpc = TransformJacobianProducts(execute_fn, gradient_fn, gradient_kwargs)
 
-    ml_boundary_execute = _get_ml_boundary_execute(interface, _grad_on_execution)
+    if interface in {"jax", "jax-python"}:
+        for tape in tapes:
+            params = tape.get_parameters(trainable_only=False)
+            tape.trainable_params = qml.math.get_trainable_indices(params)
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/execution.py:793]: More or less: I have no idea.

But this gets things working.

For torch and autograd, we can always set trainable parameters, whether or not we are using higher order derivatives.

For jax, if we the the trainable params when doing higher order derivatives, a bunch of things fail. The problem is that the jax code no longer knows the which order derivative it is registering. So we need to move that code out here instead.

File: pennylane/ops/functions/dot.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/ops/functions/dot.py:92]: So I'm actually not getting test failures when I revert this. So 🤞 it turns out I didn't actually need to do this?

File: tests/interfaces/test_jax.py (hunk 4 above)
Comment (albi3ro) [reply tests/interfaces/test_jax.py:149]: It's often more efficient to calculate the jacobian at same time as the results (`grad_on_execution=True`). But what if we aren't actually needing the derivatives? Then we did a bunch of work that never gets used.

In an ideal world, we would distinguish between "I just need the results" and "I'm doing a gradient calculation in the near future; please calculate the jacobians too".  

With jax, we can distinguish between those two cases. In `cost(x)`, we will use `_execute_wrapper`, and in `jax.grad(cost)(x)`, we will use `_execute_and_compute_jvp`. Two separate pathways.

Leads to some pretty nice speedups when we won't be needing the derivatives.

Screenshot 2023-11-29 at 4 32 57 PM

**Review by timmysilv — APPROVED** [review]
i guess a test to please codecov wouldn't hurt, but otherwise lgtm!

**Review by dwierichs — COMMENTED** [review]
Very nice work @albi3ro! :100: 
I mostly had small suggestions and a few questions.
A code comment about the JAX-specific higher-order derivatives preparation move would be nice to have I think :)

File: pennylane/interfaces/jax.py [hunk 7]
```diff
@@ -78,209 +207,93 @@ def get_jax_interface_name(tapes):
...
+        the returned tuple corresponds in order to the provided tapes.
+
     """
-    return tuple(tuple(_to_jax([r_])[0] for r_ in r) for r in res)
+    if logger.isEnabledFor(logging.DEBUG):  # pragma: no cover
+        logger.debug("Entry with (tapes=%s, execute_fn=%s, jpc=%s)", tapes, execute_fn, jpc)
+
+    parameters = tuple(tuple(t.get_parameters()) for t in tapes)
+
+    return _execute_vjp(parameters, _NonPytreeWrapper(tuple(tapes)), execute_fn, jpc)
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/jax.py:299]: you should see the version where we can use pytrees 😄 .  Hopefully we can make that change once `ParametrizedEvolution` and `ParametrizedHamiltonian` are pytrees.

File: pennylane/interfaces/jax.py [hunk 8]
```diff
@@ -12,27 +12,156 @@
...
+
+.. code-block:: python
+
+    def f_and_jvp(primals, tangents):
+        x = primals[0]
+        dx = qml.math.unwrap(tangents[0])
+        return x**2, 2*x*dx
+
+>>> jax.grad(registered_f_jvp)(jax.numpy.array(2.0))
+ValueError: Converting a JAX array to a NumPy array not supported when using the JAX JIT.
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/jax.py:79]: jit is tracing?

File: tests/interfaces/default_qubit_2_integration/test_jax_default_qubit_2.py [hunk 9]
```diff
@@ -446,7 +480,10 @@ def cost(a, U):
         res = cost(a, U)
         assert np.allclose(res, -jnp.cos(a), atol=atol_for_shots(shots), rtol=0)
 
-        jac_fn = jax.jacobian(cost)
+        if shots.has_partitioned_shots:
+            jac_fn = jax.jacobian(cost)
+        else:
+            jac_fn = jax.grad(cost)
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply tests/interfaces/default_qubit_2_integration/test_jax_default_qubit_2.py:479]: ok does look like you can use `jax.jacobian` with the device vjp if its only a single output.

File: pennylane/interfaces/jax.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/interfaces/jax.py:79]: Isn't tracing also happening when computing derivatives (VJP, JVP, Jacobian) without JIT?

File: pennylane/interfaces/execution.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/execution.py:795]: added.

File: pennylane/interfaces/jax.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/interfaces/jax.py:79]: Ah, thanks! Not quite sure why tracing implies compilation, still, but it's not a major issue.

**Review by dwierichs — APPROVED** [review]
Looks good to me. The JIT semantics question is super minor, feel free to ignore.
Amazing upgrade :tada:

File: pennylane/interfaces/jax.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/jax.py:79]: added a little clarification comment that "JIT" means "any type of tracing".

[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/interfaces/jax.py:79]: Thanks!

**Pull request merged**